File: src/ts/src/ads/internal/BannerAdHelper.ts
import {
    IMessageBridge,
    Utils,
} from "../../core";
import { MessageHelper } from "./MessageHelper";

export class BannerAdHelper {
    private readonly _bridge: IMessageBridge;
    private readonly _helper: MessageHelper;
    private _anchor: [number, number];
    private _position: [number, number];
    private _size: [number, number];
    private _visible: boolean;

    public constructor(bridge: IMessageBridge, helper: MessageHelper, size: [number, number]) {
        this._bridge = bridge;
        this._helper = helper;
        this._size = size;
        this._anchor = [0, 0];
        this._position = [0, 0];
        this._visible = false;
    }

    public get isLoaded(): boolean {
        const repsonse = this._bridge.call(this._helper.isLoaded);
        return Utils.toBool(repsonse);
    }

    public load(): void {
        this._bridge.call(this._helper.load);
    }

    private setPositionTopLeft(position: [number, number]) {
        const request = {
            [`x`]: Math.floor(position[0]),
            [`y`]: Math.floor(position[1]),
        };
        this._bridge.call(this._helper.setPosition, JSON.stringify(request));
        this._position = position;
    }

    public get anchor(): [number, number] {
        return this._anchor;
    }

    public set anchor(value: [number, number]) {
        this.setPositionTopLeft([
            this._position[0] - (value[0] - this._anchor[0]) * this._size[0],
            this._position[1] - (value[1] - this._anchor[1]) * this._size[1]]);
        this._anchor = value;
    }

    public get position(): [number, number] {
        return [
            this._position[0] + this._anchor[0] * this._size[0],
            this._position[1] + this._anchor[1] * this._size[1]];
    }

    public set position(value: [number, number]) {
        this.setPositionTopLeft([
            value[0] - this._anchor[0] * this._size[0],
            value[1] - this._anchor[1] * this._size[1]]);
    }

    public get size(): [number, number] {
        return this._size;
    }

    public set size(value: [number, number]) {
        this.setPositionTopLeft([
            this._position[0] - (value[0] - this._size[0]) * this._anchor[0],
            this._position[1] - (value[1] - this._size[1]) * this._anchor[1]]);
        const request = {
            [`width`]: Math.floor(value[0]),
            [`height`]: Math.floor(value[1]),
        };
        this._bridge.call(this._helper.setSize, JSON.stringify(request));
        this._size = value;
    }

    public get isVisible(): boolean {
        return this._visible;
    }

    public set isVisible(value: boolean) {
        this._bridge.call(this._helper.setVisible, Utils.toString(value));
        this._visible = value;
    }
}

File: src/android/ads/src/main/java/com/ee/internal/BannerAdHelper.kt
package com.ee.internal

import android.graphics.Point
import androidx.annotation.AnyThread
import com.ee.IBannerAd
import com.ee.IMessageBridge
import com.ee.Utils
import kotlinx.serialization.Serializable

/**
 * Created by Zinge on 10/12/17.
 */
class BannerAdHelper(private val _bridge: IMessageBridge,
                     private val _view: IBannerAd,
                     private val _helper: MessageHelper) {
    @Serializable
    @Suppress("unused")
    private class GetPositionResponse(
        val x: Int,
        val y: Int
    )

    @Serializable
    private class SetPositionRequest(
        val x: Int,
        val y: Int
    )

    @Serializable
    @Suppress("unused")
    private class GetSizeResponse(
        val width: Int,
        val height: Int
    )

    @Serializable
    private class SetSizeRequest(
        val width: Int,
        val height: Int
    )

    @AnyThread
    fun registerHandlers() {
        _bridge.registerHandler(_helper.isLoaded) {
            Utils.toString(_view.isLoaded)
        }
        _bridge.registerHandler(_helper.load) {
            _view.load()
            ""
        }
        _bridge.registerHandler(_helper.getPosition) {
            val position = _view.position
            val response = GetPositionResponse(position.x, position.y)
            response.serialize()
        }
        _bridge.registerHandler(_helper.setPosition) { message ->
            val request = deserialize<SetPositionRequest>(message)
            _view.position = Point(request.x, request.y)
            ""
        }
        _bridge.registerHandler(_helper.getSize) {
            val size = _view.size
            val response = GetSizeResponse(size.x, size.y)
            response.serialize()
        }
        _bridge.registerHandler(_helper.setSize) { message ->
            val request = deserialize<SetSizeRequest>(message)
            _view.size = Point(request.width, request.height)
            ""
        }
        _bridge.registerHandler(_helper.isVisible) {
            Utils.toString(_view.isVisible)
        }
        _bridge.registerHandler(_helper.setVisible) { message ->
            _view.isVisible = Utils.toBoolean(message)
            ""
        }
    }

    @AnyThread
    fun deregisterHandlers() {
        _bridge.deregisterHandler(_helper.isLoaded)
        _bridge.deregisterHandler(_helper.load)
        _bridge.deregisterHandler(_helper.getPosition)
        _bridge.deregisterHandler(_helper.setPosition)
        _bridge.deregisterHandler(_helper.getSize)
        _bridge.deregisterHandler(_helper.setSize)
        _bridge.deregisterHandler(_helper.isVisible)
        _bridge.deregisterHandler(_helper.setVisible)
    }
}